src/curves/mod.rs

```diff
@@ -1,83 +1,89 @@
+/// `y^2 = x^3 + ax + b`
 pub trait CurveParams: 'static + Copy + Clone + fmt::Debug + Default + Eq + Ord {
   /// Integer field element type
-  type FieldElement: FiniteField + Neg + Mul;
+  type BaseField: FiniteField + Neg + Mul;
+
   /// Order of this elliptic curve, i.e. number of elements in the scalar field.
   const ORDER: u32;
+
   /// Coefficient `a` in the Weierstrass equation of this elliptic curve.
-  const EQUATION_A: Self::FieldElement;
+  const EQUATION_A: Self::BaseField;
+
   /// Coefficient `b` in the Weierstrass equation of this elliptic curve.
-  const EQUATION_B: Self::FieldElement;
+  const EQUATION_B: Self::BaseField;
+
   /// Generator of this elliptic curve.
-  const GENERATOR: (Self::FieldElement, Self::FieldElement);
-  // hack: 3 and 2 to satisfy the Add<AffinePoint> trait implementation
-  const THREE: Self::FieldElement;
-  const TWO: Self::FieldElement;
-
-  // maybe Curve::uint type is diff from PCP::fieldelement type
-  // maybe:
-  // type AffinePoint;
-  // type ProjectivePoint;
-  // type Scalar;
+  const GENERATOR: (Self::BaseField, Self::BaseField);
 }
 
 /// An Affine Coordinate Point on a Weierstrass elliptic curve
 #[derive(Clone, Debug, Copy, PartialEq, Eq)]
 pub enum AffinePoint<C: CurveParams> {
-  XY(C::FieldElement, C::FieldElement),
-  Infty,
+  /// A point on the curve.
+  PointOnCurve(C::BaseField, C::BaseField),
+
+  /// The point at infinity.
+  Infinity,
 }
 
 impl<C: CurveParams> AffinePoint<C> {
-  pub fn new(x: C::FieldElement, y: C::FieldElement) -> Self {
+  /// Create a new point on the curve so long as it satisfies the curve equation.
+  ///
+  /// ## Panics
+  /// If the point is not on the curve; validated by checking if `y^2 = x^3 + ax + b`.
+  pub fn new(x: C::BaseField, y: C::BaseField) -> Self {
     // okay so this is breaking because the curve equation doesn't know how to plug in polynomials.
     // y = 31x -> y^2 = 52x^2
     // x = 36 -> x^3 = 95 + 3
     // 52x^2 = 98 ???
     assert_eq!(y * y, x * x * x + C::EQUATION_A * x + C::EQUATION_B, "Point is not on curve");
-    Self::XY(x, y)
-  }
-
-  pub fn new_infty() -> Self { Self::Infty }
-
-  pub fn negate(&mut self) {
-    match self {
-      Self::XY(_x, ref mut y) => *y = -*y,
-      Self::Infty => (),
-    }
+    Self::PointOnCurve(x, y)
   }
 }
 
 // Example:
 // Base
 
-impl<C: CurveParams> std::ops::Neg for AffinePoint<C> {
+impl<C: CurveParams> Neg for AffinePoint<C> {
   type Output = AffinePoint<C>;
 
   fn neg(self) -> Self::Output {
     let (x, y) = match self {
-      AffinePoint::XY(x, y) => (x, y),
-      AffinePoint::Infty => panic!("Cannot double point at infinity"),
+      AffinePoint::PointOnCurve(x, y) => (x, -y),
+      AffinePoint::Infinity => panic!("Cannot double point at infinity"),
     };
-    AffinePoint::new(x, C::FieldElement::ZERO - y)
+    AffinePoint::new(x, C::BaseField::ZERO - y)
   }
 }
+
+// TODO: This should likely use a `Self::ScalarField` instead of `u32`.
```

Comment:
not sure I see why? Scaler mult should work for any number (point can be doubled any amount of times), in the field or not. if we constrain the type here what does it give us? It seems like it would make the interface harder. I am down to turn it into the scalar field under the hood here, but that seems like an optimization

Reply:
My thought is that we should define a `ScalarField` for the curve anyway because we will eventually need access to this. It's a bit odd to multiply by a number that isn't in the scalar field of a curve, it's not really very intuitive to think about it that way to me at least since you should reduce `mod SCALAR_FIELD_ORDER` prior to computing it.

Reply:
As far as i understand all elliptic curve points should be able to be multiplied by an arbitrarily large number. (under the hood this is just point doubling and point addition). No it's clear that it might be redundant to do so (especially if it is bigger than the scaler field order like you point out). But i think the interface should take any number and just abstractly mod the scaler order under the hood before point doubling. That just feels like a more flexible interface to me. But also i don't mind changing it since this is for learning.